gum/arch-mips/gummipswriter.c

```diff
@@ -0,0 +1,774 @@
+
+static void
+gum_mips_writer_put_argument_list_teardown (GumMipsWriter * self,
+                                            guint n_args)
+{
+  if (n_args > 4)
+  {
+    gum_mips_writer_put_addi_reg_imm (self, MIPS_REG_SP,
+        (n_args - 4) * sizeof (guint32));
+  }
+}
+
+gboolean
+gum_mips_writer_can_branch_address (GumAddress from,
+                                    GumAddress to)
+{
+  gint64 lower_limit = (from & 0xf0000000);
+  gint64 upper_limit = (from & 0xf0000000) + GUM_INT28_MASK;
+
+  return lower_limit < to && to < upper_limit;
+}
+
+void
+gum_mips_writer_put_j_address (GumMipsWriter * self,
+                               GumAddress address)
+{
+
+  g_assert_cmpint (address & 0xf0000000, ==, self->pc & 0xf0000000);
+  g_assert_cmpint (address % 4, ==, 0);
+
+  gum_mips_writer_put_instruction (self,
+      0x08000000 | ((address & GUM_INT28_MASK) / 4));
+  gum_mips_writer_put_nop (self);
+}
+
+void
+gum_mips_writer_put_j_label (GumMipsWriter * self,
+                             gconstpointer label_id)
+{
+  gum_mips_writer_add_label_reference_here (self, label_id);
+  gum_mips_writer_put_instruction (self, 0x08000000);
+  gum_mips_writer_put_nop (self);
+}
+
+void
+gum_mips_writer_put_jr_reg (GumMipsWriter * self,
+                            mips_reg reg)
+{
+  GumMipsRegInfo ri;
+
+  gum_mips_writer_describe_reg (self, reg, &ri);
+
+  gum_mips_writer_put_instruction (self, 0x00000008 | (ri.index << 21));
+  gum_mips_writer_put_nop (self);
+}
+
+void
+gum_mips_writer_put_jal_address (GumMipsWriter * self,
+                                 guint32 address)
+{
+  gum_mips_writer_put_instruction (self, 0x0c000000 |
+      ((address & GUM_INT28_MASK) >> 2));
+  gum_mips_writer_put_nop (self);
+}
+
+void
+gum_mips_writer_put_jalr_reg (GumMipsWriter * self,
+                              mips_reg reg)
+{
+  GumMipsRegInfo ri;
+
+  gum_mips_writer_describe_reg (self, reg, &ri);
+
+  gum_mips_writer_put_instruction (self, 0x00000009 | (ri.index << 21) |
+      (GUM_MREG_RA << 11));
+  gum_mips_writer_put_nop (self);
+}
+
+void
+gum_mips_writer_put_ret (GumMipsWriter * self)
+{
+  gum_mips_writer_put_jr_reg (self, MIPS_REG_RA);
+  gum_mips_writer_put_nop (self);
+}
+
+void
+gum_mips_writer_put_la_reg_address (GumMipsWriter * self,
+                                    mips_reg reg,
+                                    GumAddress address)
+{
+  gum_mips_writer_put_lui_reg_imm (self, reg, address >> 16);
+  gum_mips_writer_put_ori_reg_reg_imm (self, reg, reg, address & 0xffff);
+}
+
+void
+gum_mips_writer_put_lui_reg_imm (GumMipsWriter * self,
+                                 mips_reg reg,
+                                 guint imm)
+{
+  GumMipsRegInfo ri;
+
+  gum_mips_writer_describe_reg (self, reg, &ri);
+
+  gum_mips_writer_put_instruction (self, 0x3c000000 | (ri.index << 16) |
+      (imm & 0xffff));
+}
+
+void
+gum_mips_writer_put_ori_reg_reg_imm (GumMipsWriter * self,
+                                     mips_reg dest_reg,
```

Comment:
We should stick to `dst_reg` here and below, as it's what we use above and in the other writers.